File: Collections Java/HashTable/TwoSum.py
# our two sum function which will return
# all pairs in the list that sum up to S
def twoSum(arr, S):

  sums = []
  hashTable = {}

  # check each element in array
  for i in range(0, len(arr)):

    # calculate S minus current element
    sumMinusElement = S - arr[i]

    # check if this number exists in hash table
    # if so then we found a pair of numbers that sum to S
    if sumMinusElement in hashTable:
      sums.append([arr[i], sumMinusElement])

    # add the current number to the hash table
    hashTable[arr[i]] = arr[i]

  # return all pairs of integers that sum to S
  return sums

print twoSum([3, 5, 2, -4, 8, 11], 7) 

File: Collections Java/HashTable/TwoSum.java
import java.util.Hashtable;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.Map;


class TwoSum{	 //all pairs in the array that sum up to S

	public static void main(String[] args) {
		int[] arr = {4, 5, 1, 8};
		int totalSum = 6;

		TwoSum obj = new TwoSum();
		obj.twoSum(arr,totalSum);
	}

	public void twoSum(int[] arr, int totalSum){
		ArrayList<Integer> numbers = new ArrayList<>();
		Hashtable<Integer,Integer> pair = new Hashtable<>();		

		for(int i=0;i<arr.length;i++){
			numbers.add(arr[i]);	
		}
		
		for(int i=0;  i < arr.length;  i++){
			int sumMinusElement =  totalSum - arr[i];	//eg 6-2 = 4 but we want total 6

			if(numbers.contains(sumMinusElement)){	//We found pair
				pair.put(arr[i], sumMinusElement);

				System.out.println("PUTTED IN HASHTABLE");
			}else{
				System.out.println("ELEMENT NOT FOUND");
			}
		}

		Enumeration<Integer> e = pair.elements();
			while(e.hasMoreElements()) {
	        	 System.out.print(e.nextElement() + ", ");
			}

		/*for(Map.Entry<Integer,Integer> m : pair.entrySet()){  
   			System.out.println(m.getKey()+" "+m.getValue());  
  		}  */	



	}
}